### PR TITLE
Dapr Error Handling/Codes - Add Internal Shared Functionality

### PR DESCRIPTION
**Pull request by robertojrojas**
# Description

This PR adds functionality to be shared between `the Dapr runtime` and various `components`.
For more details see, issue #56 and [Dapr Error Handling/Codes Proposal](https://github.com/dapr/proposals/blob/main/0009-BCIRS-error-handling-codes.md)

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #56 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests

### CODE REVIEW

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:76]: NewDaprError

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:56]: nit: just call it resourceInfo

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:44]: ResourceInfo

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:60]: EnableComponentErrorCode

Just take the map and let the runtime know if the feature is enabled.

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:66]: IsComponentErrorCodeEnabled

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:183]: Use it as an OO method of DaprError instead

**Comment by robertojrojas** [comment]
Snippet from Component side (Redis):
```golang
	if de := kitErrorCodes.NewDaprError(err, req.Metadata, kitErrorCodes.WithReason(kitErrorCodes.StateETagMismatchReason)); de != nil {
				de.SetResourceInfoData(&r.resourceInfoData)
				de.SetDescription(fmt.Sprintf("state store Set - possible etag(%s) %s. original error: %v", *req.ETag, string(state.ETagMismatch), err))
				return de
}
```

**Review by artursouza — DISMISSED** [review]

**Comment by artursouza** [comment]
LGTM but needs lint fixes.

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:69]: If metadata can be nil above, it can be nil here too.

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:36]: Can we move the reasons out of kit? I see this list growing and it would be odd to force every new error to also have a kit change.

**Review by artursouza — COMMENTED** [review]
Some small changes. It is close.

File: errorcodes/errorcodes.go
Comment (robertojrojas) [reply errorcodes/errorcodes.go:69]: Something I learned from one of the comments from other PRs about nil maps in Go is that lookup operations can be performed safely and will return the zero value of the entry value type. In this case false. Only adding an entry to a nil map is not allowed.

File: errorcodes/errorcodes.go
Comment (robertojrojas) [reply errorcodes/errorcodes.go:36]: That's a great point! Let me change it so only the `Reason` definition is in kit, but code depending on this functionality like `components-contrib` can add the various Reasons.

File: errorcodes/errorcodes.go
Comment (artursouza) [inline errorcodes/errorcodes.go:40]: Can't this be a function of DaprError instead?

**Review by artursouza — CHANGES_REQUESTED** [review]

File: errorcodes/errorcodes.go
Comment (robertojrojas) [reply errorcodes/errorcodes.go:40]: The idea of the `WithErrorReason` is to provide the ability to create a `ErrorOption` values which are used as part of the DaprError construction. This is how I envision it'll be used within components-contrib
```golang
var (
	StateETagMismatchReason = errorcodes.WithErrorReason(StateStoreEtagMismatch, 400, codes.Aborted)
	StateETagInvalidReason  = errorcodes.WithErrorReason(StateStoreETagInvalid, 400, codes.Aborted)
)
...
if de := state.NewDaprError(err, req.Metadata, state.StateETagMismatchReason); de != nil {
				de.SetResourceInfo(resourceInfo)
				de.SetDescription(fmt.Sprintf("state store Set - possible etag(%s) %s. original error: %v", *req.ETag, string(state.ETagMismatch), err))
				return de
}
```

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:1]: Consider changing the package and directory names to be `errors` to make things cleaner. Instances of the string `Error` can then be removed from names and signatures since it is implied by the package name.

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:27]: Where would this key appear?

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:30]: Do these need to be exported?

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:33]: Should this be changed to `Unknown` reason?

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:37]: Remove type stutter.

```suggestion
	Type string
	Name string
```

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:35]: Is `Info` necessary here? Can you add a comment detailing what this struct is and used for, as well as it's fields? It's not clear to me what they are for.

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:60]: Move to top of file

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:50]: type/function declaration comments should ideally start with the name of the type or function name, i.e. `// ErrorOption applies a `DaprError` to..`

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:64]: `ErrorCodesFeatureMetadataKey` should this value then have a Dapr prefix to ensure namespacing? i.e. `"_dapr_error_codes_feature"`?

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:69]: If metadata is `nil`, wont we miss setting this feature gate on?

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:75]: nit:
```suggestion
	_, ok := metadata[ErrorCodesFeatureMetadataKey]
	return ok
```

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:81]: Name implied by the repo and package name.

```suggestion
func New(err error, metadata map[string]string, options ...Option) *Error {
```

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:52]:
```suggestion
type Error struct {
```

File: go.mod
Comment (ItalyPaleAle) [inline go.mod:16]: Please use 1.56.2 which is the version currently used in the runtime, or we'll force the runtime to upgrade grpc late in the game: https://github.com/dapr/dapr/blob/ad5618711830510617e1b8fb2d0c6758f14ff6b9/go.mod#L67C3-L67C3

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:162]: What's the use case for exposing these funcs? I don't think we should let these fields be mutable after creation.

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:180]:
```suggestion
func (e *Error) ToGRPC() *status.Status {
```

File: errorcodes/errorcodes.go
Comment (ItalyPaleAle) [inline errorcodes/errorcodes.go:82]: Please add a "TODO" comment here so we can track it

File: errorcodes/errorcodes.go
Comment (ItalyPaleAle) [inline errorcodes/errorcodes.go:102]:
```suggestion
	// to override
```

File: errorcodes/errorcodes.go
Comment (ItalyPaleAle) [inline errorcodes/errorcodes.go:112]:
```suggestion
	if c != nil && c.err != nil {
```

File: errorcodes/errorcodes.go
Comment (ItalyPaleAle) [inline errorcodes/errorcodes.go:95]: Due to the check done below in `func (c *DaprError) Description() string `, we don't need this line and we can save a few bytes :)
```suggestion
```

File: errorcodes/errorcodes.go
Comment (JoshVanL) [reply errorcodes/errorcodes.go:27]: So as I understand- we are using this key to change behavior here based on a feature gate in dapr/dapr? I think it would be best for this package to be as simple as possible, and use this package or not based on the presence of the feature gate in dapr/dapr rather than needing to learn the feature gate here as well.

File: errorcodes/errorcodes.go
Comment (ItalyPaleAle) [inline errorcodes/errorcodes.go:170]: We can simplify the code and not require a method to return a gRPC status.

If you change DaprError to implement `GRPCStatus()` then it is also used by the gRPC library. See an exmaple in APIError: https://github.com/dapr/dapr/blob/ad5618711830510617e1b8fb2d0c6758f14ff6b9/pkg/messages/api_error.go#L79-L83

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:197]: Can this be pre-computed when we initialize the Error type in `New` so that we don't throw an error down stream when consumers want to get the `*status.Status`?

File: errorcodes/errorcodes.go
Comment (ItalyPaleAle) [inline errorcodes/errorcodes.go:224]: Let's work with the runtime to make it so DaprError can be returned direclty by one of the errorResponders method in pkg/http

We already have half of the work done with the errorResponseValue interface: https://github.dev/dapr/dapr/blob/ad5618711830510617e1b8fb2d0c6758f14ff6b9/pkg/http/errors.go#L20-L23

This way we can pass DaprError objects to  `respondWithError` and it will just work: https://github.dev/dapr/dapr/blob/ad5618711830510617e1b8fb2d0c6758f14ff6b9/pkg/http/responses.go#L134-L135

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:224]:
```suggestion
func (e *Errro) ToHTTP(err error) (int, []byte) {
```

File: errorcodes/errorcodes.go
Comment (JoshVanL) [inline errorcodes/errorcodes.go:234]: Same, re: pre-computing values which may throw an error at initialization.

**Review by JoshVanL — CHANGES_REQUESTED** [review]
Some comments from me!

File: errorcodes/errorcodes.go
Comment (robertojrojas) [reply errorcodes/errorcodes.go]: The `Set*` functions have been removed. The `With*` are used as optional at the time of creation.

File: errorcodes/errorcodes.go
Comment (artursouza) [reply errorcodes/errorcodes.go]: The feature flag is here because contrib also needs to know if the feature is enabled. This is not just runtime.

File: errors/errors.go
Comment (artursouza) [inline errors/errors.go:28]: This key is for internal use only as a way to pass flags between runtime and contrib. It should not need to be exported and be removed if contrib ends up not needing it.

File: errorcodes/errorcodes.go
Comment (JoshVanL) [reply errorcodes/errorcodes.go]: I think it makes more sense for all the calling code to be feature gate aware rather than kit. Making kit aware of feature flags means that its functionality is tied to the version of runtime, rather than a separate self contained utility library. It also means there would now be a bi-directional relationship between kit and other repos, where it is only one way currently.

Contrib is already tightly coupled with the runtime version, what would be the implication of having the feature gate aware code moved there instead? Another option is to have a `kit/errors.Builder` struct which is given an `Options`, however I would assume this would end up being more verbose than the caller handling the feature gate logic.

File: errors/errors.go
Comment (JoshVanL) [inline errors/errors.go:82]: Should this no be returning the `err` as is? This seems like a code smell that the caller is going to have to learn the feature gate anyway.

File: errorcodes/errorcodes.go
Comment (JoshVanL) [reply errorcodes/errorcodes.go]: @robertojrojas this comment is still relevant

File: errors/errors.go
Comment (JoshVanL) [inline errors/errors.go:132]: nit: no need to assign a variable which gets returned, and we should call the error variable `err` as this is consistent in Go. Same applies to for the other funcs below.

```suggestion
	return func(err *Error) {
		err.reason = reason
		err.grpcStatusCode = grpcStatusCode
		err.httpCode = httpCode
	}
```

File: errors/errors.go
Comment (JoshVanL) [inline errors/errors.go:202]: What is sej?

File: errors/errors.go
Comment (JoshVanL) [inline errors/errors.go:73]:
```suggestion
// New create a new Error using the supplied metadata and Options
```

File: errors/errors.go
Comment (JoshVanL) [inline errors/errors.go:49]: Please add comments explaining what these types are. As I understand it is a structured error type which helps to send the correct status codes and context over gRPC and HTTP, but that's not clear unless you read the whole file.

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: errors/errors.go
Comment (robertojrojas) [reply errors/errors.go:201]: I was using an `err` variable and I used the name to differentiate. I'm renaming to `err`.

File: errorcodes/errorcodes.go
Comment (robertojrojas) [reply errorcodes/errorcodes.go]: Agreed! I'm removing the logic of the feature gate and having it on dapr/dapr + contrib.

File: errors/errors.go
Comment (robertojrojas) [reply errors/errors.go:28]: I've decided to move this key outside of `Kit` just to keep it clean as suggested by @JoshVanL

File: errors/errors.go
Comment (JoshVanL) [inline errors/errors.go:129]:
```suggestion
func WithResourceInfo(resourceInfo *ResourceInfo) Option {
	return func(e *Error) {
		e.resourceInfo = resourceInfo
	}
```

**Review by JoshVanL — CHANGES_REQUESTED** [review]
Looks good- last couple nits and a question from me

**Review by artursouza — APPROVED** [review]

**Review by mukundansundar — APPROVED** [review]

**Review by JoshVanL — DISMISSED** [review]

File: errors/errors.go
Comment (ItalyPaleAle) [inline errors/errors.go:176]: Shouldn't this return an Internal Server Error? If I understand correctly, `WithDetails` returns an error if it fails to parse messages, which should indicate a development-time issue and not a runtime one.


```suggestion
		return status.New(codes.Internal, fmt.Sprintf("failed to create gRPC status message: %v", stErr))
```

File: errors/errors.go
Comment (ItalyPaleAle) [inline errors/errors.go:205]: Nit, for readers.

```suggestion
// JSONErrorValue implements the errorResponseValue interface (used by `github.com/dapr/dapr/pkg/http`).
```

File: errors/errors.go
Comment (ItalyPaleAle) [inline errors/errors.go:199]: Nil-check
```suggestion
    if e == nil {
        return http.StatusOK
    }
	if e.httpCode == 0 {
```

File: errors/errors.go
Comment (ItalyPaleAle) [inline errors/errors.go:209]: This is not going to work because `err.Error()` is not a JSON-encoded string.

```suggestion
        errJSON, _ := json.Marshal(fmt.Sprintf("failed to encode proto to JSON: %v", err))
		return errJSON
```

File: errors/errors.go
Comment (ItalyPaleAle) [inline errors/errors.go:191]: Same, if the expecation is that the response is JSON, the value must be JSON-encoded.

```suggestion
        errJSON, _ := json.Marshal(fmt.Sprintf("failed to encode proto to JSON: %v", err))
		return http.StatusInternalServerError, errJSON
```

File: errors/errors.go
Comment (ItalyPaleAle) [inline errors/errors.go:174]: Per the docs, `protoiface` should not be imported directly. There's an alias in the `github.com/golang/protobuf/proto` package, but that package is deprecated.

We can get around all of that (and avoid allocating a slice too, in the process) with something like:
```suggestion
	var stErr error
	ste := status.New(e.grpcStatusCode, e.description)
	if e.resourceInfo != nil {
		ste, stErr = ste.WithDetails(newErrorInfo(e.reason, e.metadata), newResourceInfo(e.resourceInfo, e.err))
	} else {
		ste, stErr = ste.WithDetails(newErrorInfo(e.reason, e.metadata))
	}
```

**Comment by ItalyPaleAle** [comment]
Thanks for your patience @robertojrojas . It looks much better, and I have only some relatively minor comments (the most critical ones are about making sure that the methods that are supposed to return JSON, always return valid JSON).

On an unrelated note, because this PR changes the `go.mod` file, we should not merge until the merge window for 1.13 opens in dapr/kit. Otherwise, importing this package in dapr/dapr for 1.12 too will cause the dependency tree to change, and it's too late in the endgame for that.

File: errors/errors.go
Comment (ItalyPaleAle) [inline errors/errors.go:225]: Please move this and the method below to a separate package (maybe `dapr/kit/grpccodes`? naming is hard - you pick)

We can then update dapr/dapr to use that package, but should be on its own

File: errors/errors.go
Comment (ItalyPaleAle) [inline errors/errors.go:25]:
```suggestion
	"google.golang.org/genproto/googleapis/rpc/errdetails"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/encoding/protojson"

	"github.com/dapr/kit/grpccodes"
```

**Review by ItalyPaleAle — APPROVED** [review]

File: grpccodes/grpccodes.go
Comment (JoshVanL) [inline grpccodes/grpccodes.go:64]: nit:

```suggestion
        default:
	  return http.StatusInternalServerError
	}
```

**Review by JoshVanL — CHANGES_REQUESTED** [review]
Couple more nits from me but otherwise LGTM

**Review by JoshVanL — APPROVED** [review]

**Review by ItalyPaleAle — APPROVED** [review]

**Pull request merged**
